File: libraries/python/src/pandaprobe/schemas.py
"""Pydantic models matching the PandaProbe backend API schemas."""

from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------


class SpanKind(StrEnum):
    AGENT = "AGENT"
    TOOL = "TOOL"
    LLM = "LLM"
    RETRIEVER = "RETRIEVER"
    CHAIN = "CHAIN"
    EMBEDDING = "EMBEDDING"
    OTHER = "OTHER"


class SpanStatusCode(StrEnum):
    UNSET = "UNSET"
    OK = "OK"
    ERROR = "ERROR"


class TraceStatus(StrEnum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    ERROR = "ERROR"


class ScoreDataType(StrEnum):
    NUMERIC = "NUMERIC"
    BOOLEAN = "BOOLEAN"
    CATEGORICAL = "CATEGORICAL"


# ---------------------------------------------------------------------------
# Span
# ---------------------------------------------------------------------------


class SpanData(BaseModel):
    model_config = {"populate_by_name": True}

    span_id: UUID = Field(default_factory=uuid4)
    parent_span_id: UUID | None = None
    name: str = Field(min_length=1, max_length=512)
    kind: SpanKind = SpanKind.OTHER
    status: SpanStatusCode = SpanStatusCode.UNSET
    input: Any | None = None
    output: Any | None = None
    model: str | None = Field(default=None, max_length=255)
    token_usage: dict[str, int] | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    started_at: datetime
    ended_at: datetime | None = None
    error: str | None = None
    completion_start_time: datetime | None = None
    model_parameters: dict[str, Any] | None = None
    cost: dict[str, float] | None = None

    def to_api_dict(self) -> dict[str, Any]:
        """Serialize to a dict matching the backend POST /traces span schema."""
        d: dict[str, Any] = {}
        d["span_id"] = str(self.span_id)
        if self.parent_span_id is not None:
            d["parent_span_id"] = str(self.parent_span_id)
        d["name"] = self.name
        d["kind"] = self.kind.value
        d["status"] = self.status.value
        if self.input is not None:
            d["input"] = _safe_json(self.input)
        if self.output is not None:
            d["output"] = _safe_json(self.output)
        if self.model is not None:
            d["model"] = self.model
        if self.token_usage is not None:
            d["token_usage"] = self.token_usage
        if self.metadata:
            d["metadata"] = self.metadata
        d["started_at"] = _iso(self.started_at)
        if self.ended_at is not None:
            d["ended_at"] = _iso(self.ended_at)
        if self.error is not None:
            d["error"] = self.error
        if self.completion_start_time is not None:
            d["completion_start_time"] = _iso(self.completion_start_time)
        if self.model_parameters is not None:
            d["model_parameters"] = self.model_parameters
        if self.cost is not None:
            d["cost"] = self.cost
        return d


# ---------------------------------------------------------------------------
# Trace
# ---------------------------------------------------------------------------


class TraceData(BaseModel):
    model_config = {"populate_by_name": True}

    trace_id: UUID = Field(default_factory=uuid4)
    name: str = Field(min_length=1, max_length=512)
    status: TraceStatus = TraceStatus.COMPLETED
    input: Any | None = None
    output: Any | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    started_at: datetime
    ended_at: datetime | None = None
    session_id: str | None = Field(default=None, max_length=255)
    user_id: str | None = Field(default=None, max_length=255)
    tags: list[str] = Field(default_factory=list)
    environment: str | None = Field(default=None, max_length=255)
    release: str | None = Field(default=None, max_length=255)
    spans: list[SpanData] = Field(default_factory=list, max_length=500)

    def to_api_dict(self) -> dict[str, Any]:
        """Serialize to a dict matching the backend POST /traces schema."""
        d: dict[str, Any] = {
            "trace_id": str(self.trace_id),
            "name": self.name,
            "status": self.status.value,
            "started_at": _iso(self.started_at),
        }
        if self.input is not None:
            d["input"] = _safe_json(self.input)
        if self.output is not None:
            d["output"] = _safe_json(self.output)
        if self.metadata:
            d["metadata"] = self.metadata
        if self.ended_at is not None:
            d["ended_at"] = _iso(self.ended_at)
        if self.session_id is not None:
            d["session_id"] = self.session_id
        if self.user_id is not None:
            d["user_id"] = self.user_id
        if self.tags:
            d["tags"] = self.tags
        if self.environment is not None:
            d["environment"] = self.environment
        if self.release is not None:
            d["release"] = self.release
        if self.spans:
            d["spans"] = [s.to_api_dict() for s in self.spans]
        return d


# ---------------------------------------------------------------------------
# Score
# ---------------------------------------------------------------------------


class ScoreData(BaseModel):
    trace_id: UUID
    name: str
    value: str
    data_type: ScoreDataType = ScoreDataType.NUMERIC
    source: str = "PROGRAMMATIC"
    reason: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    def to_api_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "trace_id": str(self.trace_id),
            "name": self.name,
            "value": self.value,
            "data_type": self.data_type.value,
            "source": self.source,
        }
        if self.reason is not None:
            d["reason"] = self.reason
        if self.metadata:
            d["metadata"] = self.metadata
        return d


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _iso(dt: datetime) -> str:
    """Format a datetime as an ISO 8601 string with timezone."""
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.isoformat()


def _safe_json(value: Any) -> Any:
    """Ensure a value is JSON-serializable; fall back to repr()."""
    if isinstance(value, (str, int, float, bool, type(None))):
        return value
    if isinstance(value, (list, tuple)):
        return [_safe_json(v) for v in value]
    if isinstance(value, dict):
        return {str(k): _safe_json(v) for k, v in value.items()}
    try:
        import json

        json.dumps(value)
        return value
    except (TypeError, ValueError):
        return repr(value)

File: libraries/typescript/src/schemas.ts
/**
 * Schema types and serializers matching the PandaProbe backend API.
 *
 * Mirrors the Python `schemas.py` Pydantic models and their `to_api_dict()`
 * serialization (omit null/empty fields, stringify UUIDs, ISO-8601 datetimes,
 * enum string values).
 */

// ---------------------------------------------------------------------------
// Enums (string-literal unions backed by const objects)
// ---------------------------------------------------------------------------

export const SpanKind = {
  AGENT: "AGENT",
  TOOL: "TOOL",
  LLM: "LLM",
  RETRIEVER: "RETRIEVER",
  CHAIN: "CHAIN",
  EMBEDDING: "EMBEDDING",
  OTHER: "OTHER",
} as const;
export type SpanKind = (typeof SpanKind)[keyof typeof SpanKind];

export const SpanStatusCode = {
  UNSET: "UNSET",
  OK: "OK",
  ERROR: "ERROR",
} as const;
export type SpanStatusCode = (typeof SpanStatusCode)[keyof typeof SpanStatusCode];

export const TraceStatus = {
  PENDING: "PENDING",
  RUNNING: "RUNNING",
  COMPLETED: "COMPLETED",
  ERROR: "ERROR",
} as const;
export type TraceStatus = (typeof TraceStatus)[keyof typeof TraceStatus];

export const ScoreDataType = {
  NUMERIC: "NUMERIC",
  BOOLEAN: "BOOLEAN",
  CATEGORICAL: "CATEGORICAL",
} as const;
export type ScoreDataType = (typeof ScoreDataType)[keyof typeof ScoreDataType];

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Format a Date as an ISO-8601 string (UTC). */
export function iso(dt: Date): string {
  return dt.toISOString();
}

/** Ensure a value is JSON-serializable; fall back to String(value). */
export function safeJson(value: unknown): unknown {
  if (value === null || value === undefined) {
    return value;
  }
  const t = typeof value;
  if (t === "string" || t === "number" || t === "boolean") {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((v) => safeJson(v));
  }
  if (t === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[String(k)] = safeJson(v);
    }
    return out;
  }
  try {
    JSON.stringify(value);
    return value;
  } catch {
    return String(value);
  }
}

/** Generate a UUID v4 string. */
export function newUuid(): string {
  return crypto.randomUUID();
}

// ---------------------------------------------------------------------------
// Span
// ---------------------------------------------------------------------------

export interface SpanDataInit {
  spanId?: string;
  parentSpanId?: string | null;
  name: string;
  kind?: SpanKind;
  status?: SpanStatusCode;
  input?: unknown;
  output?: unknown;
  model?: string | null;
  tokenUsage?: Record<string, number> | null;
  metadata?: Record<string, unknown>;
  startedAt: Date;
  endedAt?: Date | null;
  error?: string | null;
  completionStartTime?: Date | null;
  modelParameters?: Record<string, unknown> | null;
  cost?: Record<string, number> | null;
}

export class SpanData {
  spanId: string;
  parentSpanId: string | null;
  name: string;
  kind: SpanKind;
  status: SpanStatusCode;
  input: unknown;
  output: unknown;
  model: string | null;
  tokenUsage: Record<string, number> | null;
  metadata: Record<string, unknown>;
  startedAt: Date;
  endedAt: Date | null;
  error: string | null;
  completionStartTime: Date | null;
  modelParameters: Record<string, unknown> | null;
  cost: Record<string, number> | null;

  constructor(init: SpanDataInit) {
    this.spanId = init.spanId ?? newUuid();
    this.parentSpanId = init.parentSpanId ?? null;
    this.name = init.name;
    this.kind = init.kind ?? SpanKind.OTHER;
    this.status = init.status ?? SpanStatusCode.UNSET;
    this.input = init.input ?? null;
    this.output = init.output ?? null;
    this.model = init.model ?? null;
    this.tokenUsage = init.tokenUsage ?? null;
    this.metadata = init.metadata ?? {};
    this.startedAt = init.startedAt;
    this.endedAt = init.endedAt ?? null;
    this.error = init.error ?? null;
    this.completionStartTime = init.completionStartTime ?? null;
    this.modelParameters = init.modelParameters ?? null;
    this.cost = init.cost ?? null;
  }

  /** Serialize to a dict matching the backend POST /traces span schema. */
  toApiDict(): Record<string, unknown> {
    const d: Record<string, unknown> = {};
    d.span_id = this.spanId;
    if (this.parentSpanId !== null) {
      d.parent_span_id = this.parentSpanId;
    }
    d.name = this.name;
    d.kind = this.kind;
    d.status = this.status;
    if (this.input !== null && this.input !== undefined) {
      d.input = safeJson(this.input);
    }
    if (this.output !== null && this.output !== undefined) {
      d.output = safeJson(this.output);
    }
    if (this.model !== null) {
      d.model = this.model;
    }
    if (this.tokenUsage !== null) {
      d.token_usage = this.tokenUsage;
    }
    if (this.metadata && Object.keys(this.metadata).length > 0) {
      d.metadata = this.metadata;
    }
    d.started_at = iso(this.startedAt);
    if (this.endedAt !== null) {
      d.ended_at = iso(this.endedAt);
    }
    if (this.error !== null) {
      d.error = this.error;
    }
    if (this.completionStartTime !== null) {
      d.completion_start_time = iso(this.completionStartTime);
    }
    if (this.modelParameters !== null) {
      d.model_parameters = this.modelParameters;
    }
    if (this.cost !== null) {
      d.cost = this.cost;
    }
    return d;
  }
}

// ---------------------------------------------------------------------------
// Trace
// ---------------------------------------------------------------------------

export interface TraceDataInit {
  traceId?: string;
  name: string;
  status?: TraceStatus;
  input?: unknown;
  output?: unknown;
  metadata?: Record<string, unknown>;
  startedAt: Date;
  endedAt?: Date | null;
  sessionId?: string | null;
  userId?: string | null;
  tags?: string[];
  environment?: string | null;
  release?: string | null;
  spans?: SpanData[];
}

export class TraceData {
  traceId: string;
  name: string;
  status: TraceStatus;
  input: unknown;
  output: unknown;
  metadata: Record<string, unknown>;
  startedAt: Date;
  endedAt: Date | null;
  sessionId: string | null;
  userId: string | null;
  tags: string[];
  environment: string | null;
  release: string | null;
  spans: SpanData[];

  constructor(init: TraceDataInit) {
    this.traceId = init.traceId ?? newUuid();
    this.name = init.name;
    this.status = init.status ?? TraceStatus.COMPLETED;
    this.input = init.input ?? null;
    this.output = init.output ?? null;
    this.metadata = init.metadata ?? {};
    this.startedAt = init.startedAt;
    this.endedAt = init.endedAt ?? null;
    this.sessionId = init.sessionId ?? null;
    this.userId = init.userId ?? null;
    this.tags = init.tags ?? [];
    this.environment = init.environment ?? null;
    this.release = init.release ?? null;
    this.spans = init.spans ?? [];
  }

  /** Serialize to a dict matching the backend POST /traces schema. */
  toApiDict(): Record<string, unknown> {
    const d: Record<string, unknown> = {
      trace_id: this.traceId,
      name: this.name,
      status: this.status,
      started_at: iso(this.startedAt),
    };
    if (this.input !== null && this.input !== undefined) {
      d.input = safeJson(this.input);
    }
    if (this.output !== null && this.output !== undefined) {
      d.output = safeJson(this.output);
    }
    if (this.metadata && Object.keys(this.metadata).length > 0) {
      d.metadata = this.metadata;
    }
    if (this.endedAt !== null) {
      d.ended_at = iso(this.endedAt);
    }
    if (this.sessionId !== null) {
      d.session_id = this.sessionId;
    }
    if (this.userId !== null) {
      d.user_id = this.userId;
    }
    if (this.tags.length > 0) {
      d.tags = this.tags;
    }
    if (this.environment !== null) {
      d.environment = this.environment;
    }
    if (this.release !== null) {
      d.release = this.release;
    }
    if (this.spans.length > 0) {
      d.spans = this.spans.map((s) => s.toApiDict());
    }
    return d;
  }
}

// ---------------------------------------------------------------------------
// Score
// ---------------------------------------------------------------------------

export interface ScoreDataInit {
  traceId: string;
  name: string;
  value: string;
  dataType?: ScoreDataType;
  source?: string;
  reason?: string | null;
  metadata?: Record<string, unknown>;
}

export class ScoreData {
  traceId: string;
  name: string;
  value: string;
  dataType: ScoreDataType;
  source: string;
  reason: string | null;
  metadata: Record<string, unknown>;

  constructor(init: ScoreDataInit) {
    this.traceId = init.traceId;
    this.name = init.name;
    this.value = init.value;
    this.dataType = init.dataType ?? ScoreDataType.NUMERIC;
    this.source = init.source ?? "PROGRAMMATIC";
    this.reason = init.reason ?? null;
    this.metadata = init.metadata ?? {};
  }

  toApiDict(): Record<string, unknown> {
    const d: Record<string, unknown> = {
      trace_id: this.traceId,
      name: this.name,
      value: this.value,
      data_type: this.dataType,
      source: this.source,
    };
    if (this.reason !== null) {
      d.reason = this.reason;
    }
    if (this.metadata && Object.keys(this.metadata).length > 0) {
      d.metadata = this.metadata;
    }
    return d;
  }
}
